ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +579,113 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the right input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(root.getInfo().getLeftJoinExpression(), left)) {
+        // after we lift the n-way join restriction, we should be able to support FK-joins
+        // at any level in the join tree, even after we add right-deep/bushy join tree support,
+        // because a FK-join output table has the same PK as its left input table
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final Expression joinExpression,
+                                        final PlanNode node) {
+    if (!(joinExpression instanceof ColumnReferenceExp)) {
+      return true;
+    }
+    final ColumnReferenceExp simpleJoinExpression = (ColumnReferenceExp) joinExpression;
+
+    final String joinAttributeName = simpleJoinExpression.getColumnName().text();
+    final List<DataSourceNode> dataSourceNodes = node.getSourceNodes().collect(Collectors.toList());
+    final String singleAttributeKeyName;
+
+    // n-way join sub-tree (ie, not a leaf)
+    if (isInnerNode(node)) {
+      final DataSourceNode qualifiedNode;
+
+      if (simpleJoinExpression.maybeQualifier().isPresent()) {
+        final String qualifier = simpleJoinExpression.maybeQualifier().get().text();
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getDataSource().getName().text().equals(qualifier))
+            .collect(Collectors.toList());
+
+        if (allNodes.size() != 1) {
+          // double check this
+          throw new KsqlException(String.format(
+              "Join qualifier '%s' could not be resolved (either not found or not unique).",
+              qualifier
+          ));
+        }
+        qualifiedNode = allNodes.get(0);
+      } else {
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getSchema().columns().stream()
+                .map(column -> column.name().text()).collect(Collectors.toList())
+                .contains(joinAttributeName))
+            .collect(Collectors.toList());
+
+        if (allNodes.size() != 1) {
+          // double check this
+          throw new KsqlException(String.format(
+              "Join identifier '%s' could not be resolved (either not found or not unique).",
+              joinAttributeName
+          ));
+        }
+        qualifiedNode = allNodes.get(0);
+      }
+
+      // we only support joins on single attributes so the key has a single field
```

Comment:
This comment does not make sense to me. The fact that we only support joins on single attributes does not mean a user will not try to issue a join on a single attribute that is only one of multiple key columns for a source. Do we have test coverage for this case?

Reply:
I can add a test -- for this case, we have a check in the `Analyzer` and thus would never reach this point in the code.